hpke/aead.go

```diff
@@ -59,6 +153,46 @@ func (c *sealCtx) Seal(pt, aad []byte) ([]byte, error) {
 	return ct, nil
 }
 
+// Marshal serializes an HPKE sealer according to the format specified below.
+// (expressed in TLS syntax). Note that this format is not defined by the HPKE
+// standard.
+//
+// enum { sealer(0), opener(1) } HpkeRole;
+//
+// struct {
+//     HpkeKemId kem_id;   // draft-irtf-cfrg-hpke-06
+//     HpkeKdfId kdf_id;   // draft-irtf-cfrg-hpke-06
+//     HpkeAeadId aead_id; // draft-irtf-cfrg-hpke-06
+//     opaque exporter_secret<0..255>;
+//     opaque key<0..255>;
+//     opaque base_nonce<0..255>;
+//     opaque seq<0..255>;
+// } HpkeContext;
+//
+// struct {
+//   HpkeRole role = 0; // sealer
+//   HpkeContext context;
+// } HpkeSealer;
+func (c *sealCtx) Marshal() ([]byte, error) {
+	rawCtx, err := c.encdecCtx.marshal()
+	if err != nil {
+		return nil, err
+	}
+	return append([]byte{0}, rawCtx...), nil
+}
+
+// UnmarshalSealer parses an HPKE sealer.
+func UnmarshalSealer(raw []byte) (Sealer, error) {
```

Comment:
What about using BinaryMarshaler std interface.

https://golang.org/pkg/encoding/#BinaryUnmarshaler

Reply:
We could use the `BinaryMarshaler` interface, but we can't use `BinaryUnmarshaler` without additional changes. In particular, using it would require the ability to implement the following pattern:
```
var sealer sealCtx
if err := &sealer.UnmarshalBinary(rawSealer); err != nil {
    // ...
}
```
This doesn't work because `sealCtx` isn't exported. The `Sealer` interface is exported, which is why the `UnmarshalSealer` pattern works.

If you'd like I can at least have `Export` inherit `encoding.BinaryMarshaler`.

Reply:
That is true, let's keep it as is.

>  I can at least have Export inherit encoding.BinaryMarshaler

You can implement this interface without enforicing it in the Exporter interface.

Reply:
Up to you, I'm happy to do whatever!

Reply:
Choose what you think is best.

Reply:
Ack, leaving as-is.

Reply:
For future reference, I like the idea of using BinaryMarshaler and BinaryUnmarshaler interfaces, as the latter saves a malloc in some cases.